ComponentKit/Core/Scope/CKComponentScopeFrame.mm

```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
+  id identifier = oldFrame.identifier;
+  CKAssert([oldFrame.frame isKindOfClass:[CKComponentScopeFrame class]], @"Can't inject with incorrect frame class.");
+
+  auto result = _children.insert({{componentClass, identifier}, oldFrame.frame});
+
+  CKAssert(result.second, @"Scope collision: attempting to create duplicate scope %@:%@", componentClass, identifier);
+}
+
+- (_CKChildScopeFrameRet)childScopeFrameForOldIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
+{
+  for (auto &pr : _children) {
+    if (pr.second->_handle.globalIdentifier == globalIdentifier) {
+      return _CKChildScopeFrameRet{pr.first.componentClass, pr.first.identifier, pr.second};
+    }
+  }
+  return {nil, nil, 0};
+}
+
+- (BOOL)anyDescendentHasPendingUpdates:(const CKComponentStateUpdateMap &)stateUpdates
+{
+  auto it = stateUpdates.find(_handle.globalIdentifier);
+  if (it != stateUpdates.end()) {
+    return YES;
+  }
+  // Look in children
+  for (auto &pr : _children) {
```

Comment:
const auto &pr

will avoid copying, not that it matters much since they're just pairs